driver-core/src/main/com/mongodb/internal/operation/OperationReadConcernHelper.java

```diff
@@ -23,15 +23,20 @@
 import static com.mongodb.internal.connection.ReadConcernHelper.getReadConcernDocument;
 
 final class OperationReadConcernHelper {
-    static void appendReadConcernToCommand(final SessionContext sessionContext, final BsonDocument commandDocument) {
+    static void appendReadConcernToCommand(final SessionContext sessionContext, final int maxWireVersion,
+                                           final BsonDocument commandDocument) {
         notNull("commandDocument", commandDocument);
         notNull("sessionContext", sessionContext);
 
         if (sessionContext.hasActiveTransaction()) {
             return;
         }
 
-        BsonDocument readConcernDocument = getReadConcernDocument(sessionContext);
+        if (sessionContext.isSnapshot()) {
```

Comment:
As with transactions (just above), we don't append read concern at this point, instead deferring to lower-level code in CommandMessage.  This is brittle, but it works.

Reply:
Ack

Reply:
So here we simply want to omit adding `readConcern` to `commandDocument` if either `sessionContext.hasActiveTransaction` or `sessionContext.isSnapshot`. Then I cannot understand why do we add `readConcern` here at all? This is especially unclear given that `CommandMessage.getExtraElements` also adds `readConcern`. Is this because `OperationReadConcernHelper.appendReadConcernToCommand` is called only for messages that are not `OP_MSG` (i.e., for the old wire protocol, before all messages became `OP_MSG`), and `CommandMessage.getExtraElements` is called only `OP_MSG` (i.e., for the new wire protocol)?

P.S. I realize that this question is not about the PR changes anymore, but it would be useful for me to understand the situation.

Reply:
No, it's not about OP_MSG at all.  Probably easier to explain the design over Slack though.

Reply:
Jeff, thank you for explaining this matter (and beyond) to me over Zoom in details.